File: 섹션06. DFS - 경우의 수 만들기/02-02. 중복 순열 (permutations with repetition), BOJ 15656 - N과 M (7).cpp
#include <stdio.h>

#define MAX (7 + 3)

int N, M;

int num_of_cases[MAX];
int number[MAX];

void printCases()
{
	for (int i = 0; i < M; i++) printf("%d ", num_of_cases[i]);
	putchar('\n');
}

void DFS(int depth)
{
	if (depth == M)
	{
		printCases();
		return;
	}

	for (int i = 1; i <= N; i++)
	{
		num_of_cases[depth] = number[i];

		DFS(depth + 1);		
	}
}

int main()
{
	scanf("%d %d", &N, &M);

	for (int i = 1; i <= N; i++) scanf("%d", &number[i]);

	/* 정렬 */
	for (int i = 1; i <= N - 1; i++)
	{
		for (int k = i + 1; k <= N; k++)
		{
			if (number[i] > number[k])
			{
				int tmp = number[i];
				number[i] = number[k];
				number[k] = tmp;
			}
		}
	}

	DFS(0);

	return 0;
}

File: Python/섹션06. DFS - 경우의 수 만들기/02-02. 중복 순열 (permutations with repetition), BOJ 15656 - N과 M (7).py
MAX = 7 + 3

N, M = map(int, input().split())

num_of_cases = [0] * MAX
number = [0] * MAX

def print_cases():
    print(*num_of_cases[:M])

def DFS(depth):
    if depth == M:
        print_cases()
        return
    
    for i in range(1, N + 1):
        num_of_cases[depth] = number[i]
        
        DFS(depth + 1)
        
number[1:N + 1] = sorted(list(map(int, input().split())))
        
DFS(0)